db/model/Gdoc/archieToGdoc.ts

```diff
@@ -0,0 +1,241 @@
+                styleStack.push({ link: { url: node.attribs.href } })
+            } else if (tag === "br") {
+                yield { text: "\n", style: defaultStyle }
+            }
+            yield* convertCheerioNodesToTextFragments(node.children, styleStack)
+        }
+    }
+}
+
+function* lineToBatchUpdates(line: Line): Generator<docs_v1.Schema$Request> {
+    yield {
+        insertText: {
+            location: {
+                index: 1,
+            },
+            text: "\n",
+        },
+    }
+    for (const fragment of line.fragments) {
+        // Fragments will be reversed after they are generated - this is why we first
+        // emit the style change and then the text.
+        yield {
+            updateTextStyle: {
+                range: {
+                    startIndex: 1,
+                    endIndex: fragment.text.length + 1,
+                },
+                textStyle: fragment.style,
+                fields: "bold,italic,underline,strikethrough,baselineOffset,link",
+            },
+        }
+        yield {
+            insertText: {
+                location: {
+                    index: 1,
+                },
+                text: fragment.text,
+            },
+        }
+    }
+}
+
+function articleToBatchUpdates(
+    article: OwidArticleContent
+): docs_v1.Schema$Request[] {
+    const archieMlLines = [...owidArticleToArchieMLStringGenerator(article)]
+
+    const lines: Line[] = archieMlLines.map((line) => {
+        const styleStack: docs_v1.Schema$TextStyle[] = []
+        // Cheerio.load has 3 params. Cheerio 1.0.0-rc.10 which we are using is
+        // written in typescript, so the below should work without the ugly
+        // any cast. But enzyme also uses cheerio and pulls in 1.0.0-rc.3 which
+        // was not yet in typescript and thus also pulls in @types/cheerio which
+        // does not have the correct signature for load with the third param and
+        // which seems to win the battle in the TS load order. Thus the any cast
+        const $ = (cheerio as any).load(line, null, false)
+        const fragments: TextFragment[] = [
+            ...convertCheerioNodesToTextFragments(
+                $.root().contents() as Iterable<CheerioElement>,
+                styleStack
+            ),
+        ]
+        return { fragments }
+    })
+
+    const batchUpdates: docs_v1.Schema$Request[] = lines.flatMap((l) => [
+        ...lineToBatchUpdates(l),
+    ])
+
+    // The batch updates are in their logical order here. To work in one batch update
+    // we want to insert gdoc DOM spans in reverse order, so we reverse the batches here.
+    // The code above already works under the assumption that the batches will be reversed.
+    batchUpdates.reverse()
+    return batchUpdates
+}
+
+export async function createGdocAndInsertOwidArticleContent(
+    article: OwidArticleContent
+): Promise<string> {
+    const batchUpdates = articleToBatchUpdates(article)
```

Comment:
moving this call up to the parent function would make it easier to cut out the gdrive insertion part when testing manually and not wanting to go all the way to gdoc creation

Reply:
Yeah but then you can also just call `articleToBatchUpdates`, no? I kinda like that at the moment you hand it an OwidArticleContent and it does it's thing. I'd leave it like this for now but am open to moving it up in the future if it looks useful.